p2p/src/interface/event_channel.rs

```diff
@@ -0,0 +1,299 @@
+// Copyright 2020-2021 IOTA Stiftung
+// SPDX-License-Identifier: Apache-2.0
+
+use futures::{channel::mpsc, Sink, SinkExt, Stream};
+use pin_project::pin_project;
+use std::{
+    collections::VecDeque,
+    pin::Pin,
+    task::{Context, Poll, Waker},
+};
+
+/// Configure how the network should behave in the case that the capacity of the [`EventChannel`] is reached, i.g. if
+/// the channel is full. This is relevant in cases when the frequency of messages is larger than the frequency
+/// in which the [`mpsc::Receiver`][futures::channel::mpsc::Receiver] side reads from the stream.
+pub enum ChannelSinkConfig {
+    /// Block until the channel has enough capacity for the new request.
+    ///
+    /// **Note**: This pauses all network interaction and enforces back-pressure, which may be desirable if the machine
+    /// is at its limit. But it also hinders all active actions on `StrongholdP2p`, hence asynchronous methods like
+    /// [`StrongholdP2p::send_request`][super::StrongholdP2p::send_request] will be blocked as well.
+    Block,
+    /// New events will be dropped if the channel is full.
+    DropLatest,
+    /// In case that the channel is full, store new events in a ring-buffer. If the configured capacity is reached,
+    /// older events will be dropped in favor of newer ones. Send the latest events sequentially in FIFO order when
+    /// the channel has free capacity.
+    BufferLatest,
+}
+
+/// Wrapper of a [`mpsc::channel`][futures::channel::mpsc::channel] for sending events.
+/// The [`EventChannel`] cen be configure its behaviour e capacity of the channel is reached.
+#[pin_project]
+pub struct EventChannel<T> {
+    // Actual channel
+    #[pin]
+    inner: mpsc::Sender<T>,
+    // Queue for buffering the latest events if the channel is full.
+    buffer: Option<(VecDeque<T>, usize)>,
+    // Wether the `Sink` implementation of the inner channel should be used, without a buffer.
+    //
+    // This results in the `SwarmTask` blocking until `<mpsc::Sender as Sink>::send` resolves.
+    use_inner: bool,
+    // Waker from `<EventChannel as Stream>::poll_next` that is notified if a new event was added to the buffer.
+    waker: Option<Waker>,
+}
+
+impl<T> EventChannel<T> {
+    pub fn new(capacity: usize, config: ChannelSinkConfig) -> (Self, mpsc::Receiver<T>) {
```

Comment:
Since an `EventChannel<T>` can only have the values of `ChannelSinkConfig` this implementation is Ok, but I would suggest to leave room for extension: Reduce the responsibility of `new` to one, create new functions like `new_blocking(..)`, `new_drop_latest(...)`, etc...

Reply:
I personally favor having a single new method with an enum, over having a separate `new` method for each variant. Especially since the rest of the parameters (in this case only `capacity`) is the same for all variants, hence all 3 methods would have the same signature.